test/support/test_config.h

```diff
@@ -154,6 +154,12 @@
 #define _PSTL_TEST_COMPLEX_SINH_BROKEN _PSTL_TEST_COMPLEX_OP_BROKEN
 #define _PSTL_TEST_COMPLEX_TANH_BROKEN _PSTL_TEST_COMPLEX_OP_BROKEN
 
+// oneAPI DPC++ compiler 2024.2.0 and earlier is unable to eliminate a "dead" function call to an undefined function
+// within a sycl kernel which MSVC uses to allow comparisons with literal zero without warning
+#define _PSTL_TEST_COMPARISON_BROKEN                                                                                   \
+    (__cplusplus >= 202002L && _MSVC_LANG >= 202002L && _MSVC_STL_VERSION >= 143 && _MSVC_STL_UPDATE >= 202303L &&     \

```

Comment:
The rules how Microsoft defines their language version macros: `__cplusplus` https://learn.microsoft.com/en-us/cpp/build/reference/zc-cplusplus?view=msvc-170, `_MSVC_LANG` https://learn.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=msvc-170#microsoft-specific-predefined-macros. Unfortunately, our compiler does not follow these rules closely - in particular, if I remember correctly, their values are not necessarily the same with `/Zc:__cplusplus`.

Nevertheless, **both** macros are set to e.g. `C++20` only with that extra option. Therefore, generally we should check **any** of the two (i.e., ``||``) when looking for a certain version or greater, and none of the two (i.e. ``&&``) when looking for less than a certain version.

See also the definitions of ``_ONEDPL___cplusplus`` https://github.com/oneapi-src/oneDPL/blob/456bbe553b5b4fcc523875dd507f4a75d529b494/include/oneapi/dpl/internal/common_config.h#L19-L25 and of ``TEST_STD_VER`` https://github.com/oneapi-src/oneDPL/blob/456bbe553b5b4fcc523875dd507f4a75d529b494/test/support/test_macros.h#L95-L108

Reply:
Thanks for this information, I was not aware of this.  I thought I was following the example in this file, but when I flipped the directionality of the sign I did not flip the `&&` to an `||`.  Fixed.

A broader question:  Is there a reason we cannot include `test_macros.h` in this file, and make use of the unified `TEST_STD_VER` macro?  (or combine the two headers?)

Reply:
> A broader question: Is there a reason we cannot include test_macros.h in this file, and make use of the unified TEST_STD_VER macro? (or combine the two headers?)

I do not know such a reason. As far as I understand, these two files come from different places. i.e. evolved independently till recent time. I think there is a sufficient overlap to consider unification.

Setting the problem even broader, our testing harness contains a wohle lot of poorly documented macros, routines etc. which could be reused if developers knew of their existence.